### PR TITLE
Fix targets registration on Windows

### PR DESCRIPTION
**Pull request by apwojcik**
Unfortunately, automatic target registration does not work on Windows.
The linker will not add a specified DLL as a dependency if at least a symbol from the dynamic library is not used. Therefore, linking to `migrpahx_all_targets` does not link dynamic libraries on Windows as intended.

This PR changes automatic registration to runtime manual registration. Registration is done when the target library is loaded by the `register_target()` function exported from the target library. It is called from the `migraphx::make_target()` function.

### CODE REVIEW

**Comment by pfultz2** [comment]
Can we just use the `-Wl,--no-as-needed` flag instead on windows?

**Comment by pfultz2** [comment]
Also, we could make `migraphx_all_targets` a static library and then have it load the all targets before `main`:

```cpp
__attribute__((constructor)) void load_targets()
{
    #ifdef HAVE_CPU
            make_target("cpu");
    #endif
    #ifdef HAVE_GPU
            make_target("gpu");
    #endif
    #ifdef HAVE_FPGA
            make_target("fpga");
    #endif
}
```

**Comment by apwojcik** [comment]
> Can we just use the `-Wl,--no-as-needed` flag instead on windows?

It is not supported on Windows. It is only applicable to shared objects.

**Comment by apwojcik** [comment]
> Also, we could make `migraphx_all_targets` a static library and then have it load the all targets before `main`:
> 
> ```c++
> __attribute__((constructor)) void load_targets()
> {
>             make_target("ref");
>     #ifdef HAVE_CPU
>             make_target("cpu");
>     #endif
>     #ifdef HAVE_GPU
>             make_target("gpu");
>     #endif
>     #ifdef HAVE_FPGA
>             make_target("fpga");
>     #endif
> }
> ```

This will now work on Windows. DLLs are not the same as shared objects. Furthermore, MSVC/Clang on Windows does not recognize the constructor attribute because Windows does not support it. One must implement DllMain() to serve a similar purpose. However, that is more complicated because it is process-attached and thread-attached aware—the auto registration of a target creates thread-local storage, making it unavailable after exiting from a library loading thread. A DLL on Windows should generate and return an execution context to a caller to store the per-process state.

**Comment by apwojcik** [comment]
I managed to simplify the PR and limit the number of required changes.

**Comment by pfultz2** [comment]
> Furthermore, MSVC/Clang on Windows does not recognize the constructor attribute because Windows does not support it. 

Sure, we can write it using standard C++ as well:

```cpp
struct auto_load_targets
{
    auto_load_targets()
    {
        make_target("ref");
#ifdef HAVE_CPU
        make_target("cpu");
#endif
#ifdef HAVE_GPU
        make_target("gpu");
#endif
#ifdef HAVE_FPGA
        make_target("fpga");
#endif
    }
};

static auto load_targets = auto_load_targets();
```

And this does work on windows because that is how we auto register unit tests. 

> One must implement DllMain() to serve a similar purpose. 

Why would we call `DllMain`? This is not run from a dll. This is a static library being linked into an executable.

**Comment by apwojcik** [comment]
@pfultz2 it is done

File: src/register_target.cpp [hunk 1]
```diff
@@ -32,6 +32,28 @@
...
+#ifdef HAVE_GPU
+        make_target("gpu");
+#endif
+#ifdef HAVE_FPGA
+        make_target("fpga");
+#endif
+    }
+};
+[[maybe_unused]] static auto load_targets = auto_load_targets{};
+} // namespace
```
Comment (pfultz2) [inline src/register_target.cpp:54]: This needs to go into a static/object library that we only link in for the tests. This prevents lazy loading the targets.

File: src/register_target.cpp [hunk 2]
```diff
@@ -44,8 +66,6 @@ std::unordered_map<std::string, target>& target_map()
     return m;
 }
 
-void register_target_init() { (void)target_map(); }
```
Comment (pfultz2) [inline src/register_target.cpp:47]: Dont remove this, I think this was necessary to ensure targets are loaded in some cases.

File: src/include/migraphx/register_target.hpp [hunk 3]
```diff
@@ -66,9 +65,6 @@ struct register_target_action
     }
 };
 
-template <class T>
-using auto_register_target = auto_register<register_target_action, T>;
```
Comment (pfultz2) [inline src/include/migraphx/register_target.hpp:70]: Dont remove this, We still want to provide the option to register a target with CRTP.

File: test/CMakeLists.txt [hunk 4]
```diff
@@ -33,7 +33,11 @@ if(MIGRAPHX_DISABLE_LARGE_BUFFER_TESTS)
     add_compile_definitions(MIGRAPHX_DISABLE_LARGE_BUFFER_TESTS)
 endif()
 
+add_library(register_targets STATIC register_target.cpp)
+target_link_libraries(register_targets PRIVATE migraphx)
```
Comment (pfultz2) [inline test/CMakeLists.txt:37]: I think you need to link `migraphx_all_targets` here as well to get the macros defines such as `HAVE_GPU`, `HAVE_CPU`, etc.

Comment (apwojcik) [reply test/CMakeLists.txt]: Done

File: src/register_target.cpp [hunk 5]
```diff
@@ -86,16 +84,5 @@ target make_target(const std::string& name)
     return it->second;
 }
 
-std::vector<std::string> get_targets()
```
Comment (pfultz2) [inline src/register_target.cpp:89]: This function should not be removed.

Comment (apwojcik) [reply src/register_target.cpp:89]: Something is not right; I reverted that change. I will double-check.

Comment (apwojcik) [reply src/register_target.cpp]: I fixed it.

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**